[FEATURE] New TSpec for directed graphs, called WithSourceId.

hannespetur opened the pull request:
In this graph speciality the edges store reference to it's source (as well as its target) to enable functions such as getSource(graph, vertex) and inDegree(graph, vertex) to be solved in O(1) time, instead of O(n) time it would usually take, where n is the number of edges, at the cost of memory. Here's a short, self contained, compilable example (SSCCE) which demonstrates an usecase:

```cpp
#include <iostream>
#include <ctime>

#include <seqan/basic.h>
#include <seqan/sequence.h>
#include <seqan/file.h>
#include <seqan/graph_types.h>

using namespace seqan;

typedef Graph<Directed<void, WithSourceId> > TGraphWithSourceId;
typedef VertexDescriptor<TGraphWithSourceId>::Type TVertexDescriptorWithSourceId;

typedef Graph<Directed<void> > TGraphStandard;
typedef VertexDescriptor<TGraphStandard>::Type TVertexDescriptorStandard;


int main()
{
  printf("Problem: Determining the inDegree of all vertices using a graph with 100.000 vertices and 200.000 edges.\n");
  
  {
    printf(" * Using graph that stores source ids in its edges.\n");
    TGraphWithSourceId graph;
    TVertexDescriptorWithSourceId old_vertex = addVertex(graph);
    addEdge(graph, old_vertex, old_vertex);
    
    for (int num_vertexes = 1 ; num_vertexes < 100000 ; ++num_vertexes)
    {
      TVertexDescriptorWithSourceId new_vertex = addVertex(graph);
      addEdge(graph, old_vertex, new_vertex);
      addEdge(graph, (TVertexDescriptorWithSourceId) 0, new_vertex);
    }

    clock_t begin = clock();

    for (TVertexDescriptorWithSourceId vertix = 0 ; vertix < 100000 ; ++vertix)
    {
      inDegree(graph, vertix);
    }

    printf("Problem solved in %.2f seconds.\n", double(clock()-begin) / CLOCKS_PER_SEC);
  }

  {
    printf(" * Using standard graph.\n");
    TGraphStandard graph;
    TVertexDescriptorStandard old_vertex = addVertex(graph);
    addEdge(graph, old_vertex, old_vertex);
    
    for (int num_vertexes = 1 ; num_vertexes < 100000 ; ++num_vertexes)
    {
      TVertexDescriptorStandard new_vertex = addVertex(graph);
      addEdge(graph, old_vertex, new_vertex);
      addEdge(graph, (TVertexDescriptorStandard) 0, new_vertex);
    }

    clock_t begin = clock();

    for (TVertexDescriptorStandard vertix = 0 ; vertix < 100000 ; ++vertix)
    {
      inDegree(graph, vertix);
    }
    
    printf("Problem solved in %.2f seconds.\n", double(clock()-begin) / CLOCKS_PER_SEC);
  }

  return 0;
}

```

Output:
```
Problem: Determining the inDegree of all vertices using a graph with 100.000 vertices and 200.000 edges.
 * Using graph that stores source ids in its edges.
Problem solved in 0.00 seconds.
 * Using standard graph.
Problem solved in 28.57 seconds.
```

I includes in the commit the very same tests as the standard directed graph uses and they all pass.

h-2 commented:
Thanks for the PR, we will review it shortly! Some minor style issues I saw at first glance:
 * `{` should always go on the beginning of next line
 * you should never have more multiple empty lines following each other
 * please balance braces in if-else
 * please prefix `&` always by a space character

For more information, see http://seqan.readthedocs.org/en/master/StyleGuide/Cpp.html
I know not all of these things are followed consistently throughout the codebase, but we are trying to improve it, so all new code has to fulfill it :)

h-2 commented on include/seqan/graph_types/graph_impl_directed.h:
```diff
@@ -139,6 +194,53 @@ _getEdgeIdManager(Graph<Directed<TCargo, TSpec> > const& g) {
+_copyGraph(Graph<Directed<TCargo, WithSourceId> > const& source,
+           Graph<Directed<TCargo, WithSourceId> >& dest,
+           bool transpose)
+{
```
in SeqAn the argument order is always `(dest1, dest2, src1, src2, option, tag)`, please adapt this even if the function is "private".

hannespetur commented:
Thanks for the response! I've fixed all coding style issues I could find in implementation for the directed graphs and change the arguments for `_copyGraph`. Also I changed arguments with `const` to `const &` in multiple functions, as it is usually preferred.

rrahn commented on include/seqan/graph_types/graph_impl_directed.h:
```diff
@@ -74,12 +74,13 @@ class Graph<Directed<TCargo, TSpec> >
 
-
-        Graph() {
+        Graph()
```
Please replace tabs with 4 spaces. Here you can find some notes about setting up you editor: http://seqan.readthedocs.org/en/master/HowTo/EditorSetup.html?highlight=editor

hannespetur replied:
I'm not using tabs here. Was this note supposed to be elsewhere?

rrahn commented on include/seqan/graph_types/graph_impl_directed.h:
```diff
@@ -88,19 +89,69 @@ class Graph<Directed<TCargo, TSpec> >
+
+        Graph()
+        {
+            SEQAN_CHECKPOINT
```
Please remove all instances of SEQAN_CHECKPOINT (only in the new code). There are not used anymore.

rrahn commented on include/seqan/graph_types/graph_impl_directed.h:
```diff
@@ -108,20 +159,34 @@ class Graph<Directed<TCargo, TSpec> >
+    SEQAN_CHECKPOINT
+    typedef Graph<Directed<TCargo, WithSourceId> > TGraph;
+    typedef typename EdgeType<TGraph>::Type TEdgeStump;
+    return const_cast<String<TEdgeStump*>&>(g.data_vertex_in);
```
This is a very evil thing to do. I know the old code did not look better, but never cast-away the const type of a return value, which would allow me to literally change the state of a member of a const-object. If such thing would be intended by design, one would use the mutable keyword instead. I know this is a private function, but is there any instance where this function is called on both non-const and const graphs? If not, please return it as const reference. Otherwise add a second interface with non-const graph.

rrahn commented on include/seqan/graph_types/graph_impl_directed.h:
```diff
@@ -108,20 +159,34 @@ class Graph<Directed<TCargo, TSpec> >
+template <typename TCargo, typename TSpec>
+inline typename VertexIdHandler<Graph<Directed<TCargo, TSpec> > >::Type &
+_getVertexIdManager(Graph<Directed<TCargo, TSpec> > const & g)
+{
```
Since you touched the evil interfaces :), could you please adapt them as described above.
Many thanks.

rrahn commented on include/seqan/graph_types/graph_impl_directed.h:
```diff
@@ -130,23 +195,89 @@ _getVertexIdManager(Graph<Directed<TCargo, TSpec> > const& g) {
+            else
+            {
+                _createVertices(dest,targetVertex);
+            }
```
Please, remove the curly brackets for simple if-else statements with only one statement, as long as the scope is clear from the context (no nested if statements, all if-else if-else clauses capture only one line)

rrahn commented on include/seqan/graph_types/graph_impl_directed.h:
```diff
@@ -159,22 +290,39 @@ _copyGraph(Graph<Directed<TCargo, TSpec> > const& source,
+            else
+            {
+                _createVertices(dest,targetVertex);
+            }
```
Same here.

rrahn commented on include/seqan/graph_types/graph_impl_directed.h:
```diff
@@ -159,22 +290,39 @@ _copyGraph(Graph<Directed<TCargo, TSpec> > const& source,
+            else
+            {
+                e = addEdge(dest, targetVertex, sourceVertex);
+            }
```
same here

rrahn commented on include/seqan/graph_types/graph_impl_directed.h:
```diff
@@ -334,10 +520,14 @@ inDegree(Graph<Directed<TCargo, TSpec> > const& g,
-            if ( (TVertexDescriptor) getTarget(current) == vertex) ++count;
+        while (current!=0)
+        {
+            if ((TVertexDescriptor) getTarget(current) == vertex)
```
Please, use static_cast instead of the c-style cast.

rrahn commented on include/seqan/graph_types/graph_impl_directed.h:
```diff
@@ -346,53 +536,120 @@ inDegree(Graph<Directed<TCargo, TSpec> > const& g,
 {
     SEQAN_CHECKPOINT
-    return (inDegree(g,vertex)+outDegree(g,vertex));
+    return inDegree(g,vertex)+outDegree(g,vertex);
```
please put spaces around the +-operator.

rrahn commented on include/seqan/graph_types/graph_impl_directed.h:
```diff
@@ -346,53 +536,120 @@ inDegree(Graph<Directed<TCargo, TSpec> > const& g,
+    else
+    {
+        g.data_vertex[vd] = (TEdgeStump*) 0;
+    }
```
Remove brackets

rrahn commented on include/seqan/graph_types/graph_impl_directed.h:
```diff
@@ -449,18 +757,28 @@ removeEdge(Graph<Directed<TCargo, TSpec> >& g,
-        if ( (TVertexDescriptor) getTarget(current) == target) break;
+    while (current != (TEdgeStump*) 0)
+    {
+        if ((TVertexDescriptor) getTarget(current) == target)
```
static_cast<>

rrahn commented on include/seqan/graph_types/graph_impl_directed.h:
```diff
@@ -449,18 +757,28 @@ removeEdge(Graph<Directed<TCargo, TSpec> >& g,
+    else
+    {
+        g.data_vertex[source] = getNextT(current);
+    }
```
Remove brackets

rrahn commented on include/seqan/graph_types/graph_impl_directed.h:
```diff
@@ -485,18 +803,28 @@ removeEdge(Graph<Directed<TCargo, TSpec> >& g,
+    else
+    {
+        g.data_vertex[sourceVertex(g,edge)] = getNextT(current);
+    }
```
Remove brackets

rrahn commented on include/seqan/graph_types/graph_impl_directed.h:
```diff
@@ -538,27 +867,31 @@ removeInEdges(Graph<Directed<TCargo, TSpec> >& g,
-            if ( (TVertexDescriptor) current->data_target==v) {
+        while (current!=0)
+        {
+            if ((TVertexDescriptor) current->data_target==v)
```
static_cast

rrahn commented on include/seqan/graph_types/graph_impl_directed.h:
```diff
@@ -663,32 +1015,38 @@ write(TFile & target,
     write(target, "Adjacency list:\n");
-    for(;it!=itEnd;++it, ++pos) {
-        if (!idInUse(_getVertexIdManager(g), pos)) continue;
+    for (; it!=itEnd; ++it, ++pos)
```
Spaces around !=

rrahn commented on include/seqan/graph_types/graph_impl_directed.h:
```diff
@@ -663,32 +1015,38 @@ write(TFile & target,
     pos = 0;
     write(target, "Edge list:\n");
-    for(;it!=itEnd;++it, ++pos) {
+    for (; it!=itEnd; ++it, ++pos)
```
Spaces

rrahn commented on include/seqan/graph_types/graph_impl_directed.h:
```diff
@@ -663,32 +1015,38 @@ write(TFile & target,
+        {
             write(target, "Source: ");
-            appendNumber(target, (int)pos);
+            appendNumber(target, (int) pos);
```
static_cast

rrahn commented on include/seqan/graph_types/graph_impl_directed.h:
```diff
@@ -663,32 +1015,38 @@ write(TFile & target,
+            appendNumber(target, (int) pos);
             write(target, ",Target: ");
-            appendNumber(target, (int)getTarget(current));
+            appendNumber(target, (int) getTarget(current));
```
static_cast

rrahn commented on include/seqan/graph_types/graph_impl_directed.h:
```diff
@@ -663,32 +1015,38 @@ write(TFile & target,
+            appendNumber(target, (int) getTarget(current));
             write(target, " (Id: ");
-            appendNumber(target, (int)_getId(current));
+            appendNumber(target, (int) _getId(current));
```
static_cast

rrahn commented on include/seqan/graph_types/graph_interface.h:
```diff
@@ -194,6 +193,20 @@ struct EdgeType<Graph<Tree<TCargo, TSpec> > const> {
 //////////////////////////////////////////////////////////////////////////////
 
+template<typename TCargo>
+struct EdgeType<Graph<Directed<TCargo, WithSourceId> > > {
```
Bracket on new line.

rrahn commented on include/seqan/graph_types/graph_interface.h:
```diff
@@ -194,6 +193,20 @@ struct EdgeType<Graph<Tree<TCargo, TSpec> > const> {
+//////////////////////////////////////////////////////////////////////////////
+
+template<typename TCargo>
+struct EdgeType<Graph<Directed<TCargo, WithSourceId> > const> {
```
Bracket on new line

rrahn commented:
Ok it looks good to me. If you could address the remaining minor issues, then we could merge this PR.
Thank you very much for contributing :+1:

hannespetur commented on include/seqan/graph_types/graph_impl_directed.h:
```diff
@@ -74,55 +74,112 @@ class Graph<Directed<TCargo, TSpec> >
-    SEQAN_CHECKPOINT
+template <typename TCargo, typename TSpec>
+inline String<typename EdgeType<Graph<Directed<TCargo, TSpec> > >::Type*> &
+_getVertexString(Graph<Directed<TCargo, TSpec> > const & g)
```
I decided to leave this function like it was before (except some formatting and remove SEQAN_CHECKOUT) and remove the one I create one for the in vertices. _copyGraph no longer uses this function but it is still used in the test suite.

hannespetur commented on include/seqan/graph_types/graph_impl_directed.h:
```diff
@@ -74,55 +74,112 @@ class Graph<Directed<TCargo, TSpec> >
+
+template <typename TCargo, typename TSpec>
+inline typename Size<Graph<Directed<TCargo, WithSourceId> > >::Type
+_getVertexStringSize(Graph<Directed<TCargo, TSpec> > const & g)
```
_copyGraph now uses this function instead.

rrahn replied:
As far as I can see this function is only used in _copyGraph, which is overloaded for DirectedGraphs and your specialisation?
Why bother to add a private function that returns just the size. Just call length(g.data_vertex_in) and length(g.data_vertex), where you need it.

hannespetur commented:
test_bam_io is also failing locally. Is it something I need to look into?

rrahn commented on include/seqan/graph_types/graph_impl_directed.h:
```diff
@@ -156,25 +270,35 @@ _copyGraph(Graph<Directed<TCargo, TSpec> > const& source,
     typedef typename Iterator<String<TEdgeStump*>, Standard>::Type TIter;
     clear(dest);
-    resize(dest.data_vertex, length(_getVertexString(source)));
+    resize(dest.data_vertex, _getVertexStringSize(source));
```
Just call length(source.data_vertex) here. Like the begin and end below.

rrahn commented on include/seqan/graph_types/graph_impl_directed.h:
```diff
@@ -639,8 +905,11 @@ findEdge(Graph<Directed<TCargo, TSpec> > const& g,
-        if ( (TVertexDescriptor) getTarget(current) == w) return current;
+    while (current != (TEdgeStump*) 0)
+    {
+        if ( (TVertexDescriptor) getTarget(current) == w)
```
static_cast<>

rrahn commented on include/seqan/graph_types/graph_impl_directed.h:
```diff
@@ -663,32 +931,38 @@ write(TFile & target,
-            appendNumber(target, (int)getTarget(current));
+        while (current != 0)
+        {
+            appendNumber(target, (int) getTarget(current));
```
static_cast<>

rrahn commented:
Can you please fix the last three open issues?
Then we can merge it. Btw, test_bam_io is not related to your work. So don't worry about it.

hannespetur commented:
Yes, it should be in order now. Thank you for your patience.

rrahn commented:
lgtm! Thank you again for your contribution.
